### PR TITLE
Fix #12774: skip consumer POM re-attachment on repeated task segments

### LINKED ISSUE #12774: `mvn package compiler:help` cause runtime issue
### Affected version

4.0.0-rc-6

### Bug description

found while testing https://github.com/apache/maven-jar-plugin/pull/508#issuecomment-5352213156 but clearly purely Maven core issue

on any simple Maven build with Maven 4.0.0-rc6, `mvn package compiler:help` makes maven-jar-plugin doing twice attach, and just before `compiler:help` we get a new second build header:

```
❯ mvn -V -DskipTests package compiler:help
Apache Maven 4.0.0-rc-6 (6a8189b24518daa120539fa41ce12f2b48ec09a8)
Maven home: /home/herve/.sdkman/candidates/maven/4.0.0-rc-6
Java version: 21.0.1, vendor: Oracle Corporation, runtime: /home/herve/.sdkman/candidates/java/21.0.1-open
Default locale: fr_FR, platform encoding: UTF-8, time zone: Europe/Paris
OS name: "linux", version: "6.8.0-137-generic", arch: "amd64", family: "unix"
[INFO] Scanning for projects...
[INFO] Loaded 23569 auto-discovered prefixes for remote repository central (prefixes-central.txt)
[INFO] 
[INFO] ----------------------------------< org.apache.maven.shared:maven-dependency-analyzer >-----------------------------------
[INFO] Building Apache Maven Dependency Analyzer 1.17.2-SNAPSHOT
[INFO]   from pom.xml
[INFO] ---------------------------------------------------------[ jar ]----------------------------------------------------------
...
[INFO] --- surefire:3.5.6:test (default-test) @ maven-dependency-analyzer ---
[INFO] Tests are skipped.
[INFO] 
[INFO] --- jar:3.5.0:jar (default-jar) @ maven-dependency-analyzer ---
[WARNING] artifact 'org.apache.maven.shared:maven-dependency-analyzer:pom:consumer:1.17.2-SNAPSHOT' already attached, replacing previous instance
[INFO] 
[INFO] ----------------------------------< org.apache.maven.shared:maven-dependency-analyzer >-----------------------------------
[INFO] Building Apache Maven Dependency Analyzer 1.17.2-SNAPSHOT
[INFO]   from pom.xml
[INFO] ---------------------------------------------------------[ jar ]----------------------------------------------------------
[INFO] 
[INFO] --- compiler:3.15.0:help (default-cli) @ maven-dependency-analyzer ---
...
```

### PR DESCRIPTION
**Pull request by gnodet**
## Summary

- Make `ConsumerPomArtifactTransformer.injectTransformedArtifacts()` idempotent by checking if a consumer POM artifact (classifier=`consumer`, type=`pom`) is already attached before calling `addAttachedArtifact()`
- Prevents the duplicate "Building ..." header and `artifact '...:pom:consumer:...' already attached, replacing previous instance` warning when running multiple task segments (e.g. `mvn package compiler:help`)
- Add unit test verifying that a second invocation does not duplicate the attached artifact

Fixes #12774

## Test plan

- [x] New unit test `injectTransformedArtifactsTwiceShouldNotDuplicate` verifies idempotency
- [ ] Manual verification: `mvn package compiler:help` no longer produces the warning

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: impl/maven-core/src/main/java/org/apache/maven/internal/transformation/impl/ConsumerPomArtifactTransformer.java [hunk 1]
```diff
@@ -89,7 +89,11 @@ public void injectTransformedArtifacts(RepositorySystemSession session, MavenPro
                     : Files.createTempFile(CONSUMER_POM_CLASSIFIER + "-", ".pom");
             deferDeleteFile(consumer);
 
-            project.addAttachedArtifact(createConsumerPomArtifact(project, consumer, session));
+            boolean alreadyAttached = project.getAttachedArtifacts().stream()
```
Comment (gnodet) [inline impl/maven-core/src/main/java/org/apache/maven/internal/transformation/impl/ConsumerPomArtifactTransformer.java:82]: **Low:** The idempotency check here is placed after the temp file creation and `deferDeleteFile()` registration (lines 87-90). On repeated invocations where the consumer POM is already attached, a new temp file is created on disk, registered for deferred deletion, but never used.

Consider hoisting the `alreadyAttached` check to the top of the `if (Features.consumerPom(...))` block, before creating the temp file. This would avoid unnecessary I/O on repeated invocations. The files are cleaned up by `@PreDestroy doDeleteFiles()` so this is not a leak, just wasteful work.

**Review by gnodet — COMMENTED** [review]
Correct and minimal fix that makes `injectTransformedArtifacts()` idempotent by guarding against duplicate consumer POM attachment.

One low-severity optimization opportunity noted inline.

**Positive observations:**
- `TransformedArtifact` already handles POM-change detection lazily via SHA1 checksumming in `mayUpdate()`, so skipping re-attachment is safe — the already-attached artifact will still pick up source POM changes on the next `getFile()` call.
- The matching criteria (`getClassifier()` + `getType()`) on Maven's `Artifact` interface is correct for this context.

_This review was generated by an AI agent and may contain inaccuracies. Please verify all suggestions before applying._

_Claude Code on behalf of Guillaume Nodet_

File: impl/maven-core/src/main/java/org/apache/maven/internal/transformation/impl/ConsumerPomArtifactTransformer.java (hunk 1 above)
Comment (Bukama) [reply impl/maven-core/src/main/java/org/apache/maven/internal/transformation/impl/ConsumerPomArtifactTransformer.java:82]: Do you want to apply this improvement before merging as we had some performance issue with local repository in rc6 or do you want to apply it in a separate PR/later version?

**Review by Bukama — APPROVED** [review]

**Review by gnodet — COMMENTED** [review]
**Delta review** — checking new commit `9515783` ("Hoist alreadyAttached check before temp file creation").

✅ **Previous finding resolved:** The `alreadyAttached` check is now the first statement inside the `if (Features.consumerPom(...))` block, and all temp file creation / `deferDeleteFile()` / `addAttachedArtifact` calls are wrapped inside `if (!alreadyAttached)`. No unnecessary I/O on repeated invocations.

A new regression test `injectTransformedArtifactsTwiceShouldNotDuplicate` verifies the idempotency behavior — clean, well-structured test.

No new issues introduced.

_This review was generated by an AI agent and may contain inaccuracies. Please verify all suggestions before applying._

_Claude Code on behalf of Guillaume Nodet_
